### PR TITLE
Feat/hakime doosti

### PR DESCRIPTION
**Pull request by HkDoosti**
PipelineBuilder

### CODE REVIEW

**Comment by HkDoosti** [comment]
PipelineDesignPattern

File: src/HakimeDoosti/PipelineDesignPattern/PipelineDesignPattern/CustomException.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace PipelineDesignPattern
```
Comment (mohammadKarimi) [inline src/HakimeDoosti/PipelineDesignPattern/PipelineDesignPattern/CustomException.cs:7]: Please use File scoped namespace.

`namespace PipelineDesignPattern;`

File: src/HakimeDoosti/PipelineDesignPattern/PipelineDesignPattern/Framework.cs [hunk 2]
```diff
@@ -0,0 +1,88 @@
+﻿using Dumpify;
+using System.Reflection;
+
+namespace PipelineDesignPattern
+{
+    public class Framework
+    {
+        public delegate void Action(HttpContext context);
```
Comment (mohammadKarimi) [inline src/HakimeDoosti/PipelineDesignPattern/PipelineDesignPattern/Framework.cs:8]: you should remove this delegate, and use Action(in dotnet) instead.

File: src/HakimeDoosti/PipelineDesignPattern/PipelineDesignPattern/Framework.cs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        public void ExceptionHandling(HttpContext context, Action<HttpContext> action)
+        {
+            try
+            {
+                action(context);
+            }
+            catch (CustomException ex)
```
Comment (mohammadKarimi) [inline src/HakimeDoosti/PipelineDesignPattern/PipelineDesignPattern/Framework.cs:15]: perfect 👍

File: src/HakimeDoosti/PipelineDesignPattern/PipelineDesignPattern/Framework.cs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+            if (context.IP is null) throw new CustomException("IP is null ");
+            action(context);
+
+        }
+        public void Authentication(HttpContext context, Action<HttpContext> action)
+        {
+            "begin Authentication".Dump();
+            if (context is null) throw new CustomException("context is null ");
+            if (context.IP is null) throw new CustomException("IP is null ");
+            if (context.IP == "192.15.0.0") throw new CustomException("you are from Iran");
```
Comment (mohammadKarimi) [inline src/HakimeDoosti/PipelineDesignPattern/PipelineDesignPattern/Framework.cs:40]: for equality, there are many ways. 
but for constant checking, please follow to use constant pattern matching like :

`if (context.IP is "192.15.0.0")`

File: src/HakimeDoosti/PipelineDesignPattern/PipelineDesignPattern/Framework.cs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+            action(context);
+            "end Authentication".Dump();
+
+
+        }
+
+        public void EndpointHandling(HttpContext context, Action<HttpContext> action)
+        {
+            var urlParts = context.Url.Split('/');
+            var methodName = "";
```
Comment (mohammadKarimi) [inline src/HakimeDoosti/PipelineDesignPattern/PipelineDesignPattern/Framework.cs:52]: it is more readable to use string.Empty. 
(optional)

File: src/HakimeDoosti/PipelineDesignPattern/PipelineDesignPattern/Framework.cs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+            var controllerName = "";
+            var UserId = "";
```
Comment (mohammadKarimi) [inline src/HakimeDoosti/PipelineDesignPattern/PipelineDesignPattern/Framework.cs:54]: local vatiable should be write in camel formta. 
`userId`

File: src/HakimeDoosti/PipelineDesignPattern/PipelineDesignPattern/PipelineBuilder.cs
Comment (mohammadKarimi) [inline src/HakimeDoosti/PipelineDesignPattern/PipelineDesignPattern/PipelineBuilder.cs:1]: Perfect 👍

**Review by mohammadKarimi — APPROVED** [review]
Perfect. Perfect.

really I was surprised while I was checking your code. I think, your implementation can be the best one in this repository. 

I sent you a few issues, consider that, all of them were not critical. 
hence, I will merge your code into main branch. if you want to fix them, please fix in main branch or in your new practice. 
🥇

**Pull request merged**